httpclient5-testing/src/test/java/org/apache/hc/client5/testing/async/TestConnectionClosureRace.java

```diff
@@ -0,0 +1,448 @@
+    final AtomicInteger connectionsEstablished = new AtomicInteger(0);
+    final String scheme;
+    final HttpVersionPolicy httpVersionPolicy;
+
+    volatile ServerSocket serverSocket;
+    volatile int port;
+    volatile Thread serverThread;
+
+    AbstractTestConnectionClosureRace(final String scheme, final HttpVersionPolicy httpVersionPolicy) {
+        this.scheme = scheme;
+        this.httpVersionPolicy = httpVersionPolicy;
+    }
+
+    @BeforeAll
+    static void newline() {
+        System.out.println();
+    }
+
+    @BeforeEach
+    void setup() throws Exception {
+        if ("http".equals(scheme)) {
+            serverSocket = ServerSocketFactory.getDefault().createServerSocket(0);
+        } else {
+            final SSLContext serverSSLContext = SSLTestContexts.createServerSSLContext();
+            final SSLServerSocketFactory sslServerSocketFactory = serverSSLContext.getServerSocketFactory();
+            serverSocket = sslServerSocketFactory.createServerSocket(0);
+        }
+        port = serverSocket.getLocalPort();
+
+        serverThread = new Thread(this::runServer);
+        serverThread.setDaemon(true);
+        serverThread.start();
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        if (serverSocket != null) {
+            serverSocket.close();
+        }
+        if (serverThread != null) {
+            serverThread.interrupt();
+            serverThread.join(1000);
+        }
+    }
+
+    @Test
+    @Timeout(5)
+    @Order(1)
+    void smokeTest() throws Exception {
+        try (final CloseableHttpAsyncClient client = asyncClient(false)) {
+            final SimpleHttpResponse simpleHttpResponse = sendPing(client).get();
+            assertEquals(200, simpleHttpResponse.getCode());
+        }
+    }
+
+    @ParameterizedTest(name = "Validation: {0}")
+    @ValueSource(booleans = { false, true })
+    @Timeout(5)
+    @Order(2)
+    void testSlowSequentialRequests(final boolean validateConnections) throws Exception {
+        try (final CloseableHttpAsyncClient client = asyncClient(validateConnections)) {
+            final List<Future<SimpleHttpResponse>> futures = new ArrayList<>();
+            for (int i = 0; i < 10; i++) {
+                futures.add(sendPing(client));
+                Thread.sleep(25);
+            }
+
+            checkResults(getValidationPrefix(validateConnections) + "Sequential requests (slow)", futures);
+        }
+    }
+
+    @ParameterizedTest(name = "Validation: {0}")
+    @ValueSource(booleans = { false, true })
+    @Timeout(10)
+    @Order(3)
+    void testRapidSequentialRequests(final boolean validateConnections) throws Exception {
+        try (final CloseableHttpAsyncClient client = asyncClient(validateConnections)) {
+            final List<Future<SimpleHttpResponse>> futures = new ArrayList<>();
+            for (int i = 0; i < 2500; i++) {
+                final Future<SimpleHttpResponse> f = sendPing(client);
```

Comment:
@rschmitt If you want my OK to merge this change-set please help me understand why do 
```
                final Future<SimpleHttpResponse> f = sendPing(client);
                try {
                    f.get();
                } catch (final ExecutionException ignore) {
                }
                futures.add(f);
``` 
in this instance and do 
```
                futures.add(sendPing(client));
```
in other cases?